# 0. შესავალი

ალბათ ყველას გყავთ (თუ თქვენ თვითონ არ ხართ) ისეთი მეგობარი, რომელიც ცოტა გამაღიზიანებელი სიმარტივით ხსნის ტექნიკურ პრობლემებს. თუ არ გყავთ ასეთი მეგობარი და არც თვითონ ხართ არ იდარდოთ, რადგან ამ წიგნის ბოლოს დაუმეგორდებით მონაცემთა ანალიტიკის და „მანქანური დასწავლის“ ალგორითმებსა და ხელსაწყოებს, რომლებიც იმაზე უფრო სწრაფად და უფრო რთულ პრობლემებს გადაგაჭრევინებთ ვიდრე წარმოგიდგენიათ.
 
მონაცემები დღევანდელი, ციფრული, სამყაროს განუყოფელი ნაწილია და მათი დამუშავების უნარი ყოველ წელს უფრო და უფრო ღირებული ხდება. მსგავსი კომპეტენციის ადამიანი კი უკლებლივ ყველა ზომის და ინდუსტრუიის კომპანიისთვის საგანძურს წარმოადგენს.
 
სულ რაღაც 5-10 წლის წინ, ორგანიზაციებს უწევდათ დაექირავებინათ სტატისტიკოსები, მათემატიკოსები, კომპიუტერული მეცნიერებების სპეციალისტები და სხვა ტექნიკური უნარების მქონე ადამიანები ძალიან მარტივი ამოცანების გადასაჭრელადაც კი. თუმცა ერთის მხრივ მონაცემთა ანალიზზე მასშტაბურმა მოთხოვნამ და მეორეს მხრივ ტექნოლოგიების მკვერთმა განვითარებამ შედეგად მოგვცა უამრავი მზა ხელსაწყო, რომელთა ათვისებაც არ მოითხოვს დიპლომს სარაკეტო მეცნიერებში. პირიქითაც შეგვიძლია ვთქვათ, დღეს მონაცემების დამუშავება იმდენად ინტეგრირებულია ყველა ინდუსტრიაში, რომ თავიანთი საქმის ექსპერტები შეიარაღებულნი მანქანური დასწავლის და მონაცემთა დამუშავების მეთოდებით  უმეტეს შემთხვევაში გაცილებით მეტ ღირებულებას ქმნიან, ვიდრე მხოლოდ ტექნიკური სპეციალისტები.
 
რა თქმა უნდა არ მოგატყუებთ და არ გეტვით, რომ გიგანტი ორგანიზაციები როგორებიცაა მაგალითად გუგლი, ფეისბუქი, ამაზონი, ფეიფალი და ა.შ. აღარ საჭიროებენ ისეთი კომპეტენციის კადრებს, ვინც საფუძვლიანად ფლობენ მსგავსი ტიპის ალგორითმებს. თუმცა აღნიშნული კომპანიების შემთხვევაშიც კი ხშირად იმაზე გაცილებით მარტივად იჭრება ძვირადღირებული ამოცანები ვიდრე ჩვენ წარმოგვიდგენია. 
 
წიგნი განკუთვნილია ყველა იმ ადამიანისთვის ვინც დაინტერესებულია შეიძინოს ანალიტიკური უნარები და თვითონვე, დამოუკიდებლად, შეძლოს მასთან დაკავშირებული ამოცანების გადაჭრა. თუ თქვენ აქამდე ისეთ გავრცელებულ პროგრამასთანაც კი არ გქონიათ შეხება, როგოროცაა „ექსელი“, არ იდარდოთ სწორ ადგილას მოხვდით.


# 1. მონაცემთა ანალიზი და მისი გამოყენება პრაქტიკაში

### 1.1 მონაცემთა ანალიზის მნიშვნელობა

Სანამ ანალიზის ტიპებსა და ალგორითმების ახსნას შევეჭიდებით, Მოგიყვებით ერთ ძალიან საინტერესო ისტორიას, რომელიც ხაზს უსვამს მონაცემთა ანალიტიკის მნიშვნელობას. ეს ისოტორია ცნობილია როგორც აბრაჰამ ვალდის „გადარჩენილების მიკერძოებულობა.“

ვალდი 1902 წლის 31 ოქტომბერს, უნგრეთის ქალაქ კოლოსვარში (ახლანდელი კლუჟი, რუმინეთი) ებრაულ ოჯახში დაიბადა. უნგრეთში განათლების მიღება გაუჭირდა, რადგან სწავლა შაბათობით იყო, ხოლო ებრაული ტრადიციების შესაბამისად ეს დასვენების დღეა. თუმცა გამორჩეული ნიჭის მქონე ვალდი ადრეულ ასაკშივე, ვენის უნივერსიტეტში მოხვდა. 1938 წლამდე (სანამ ნაცისტური გერმანია შეიჭრებოდა ვენაში) ავსტიაში ცხოვრება შედარებით უკეთესი იყო. მას შემდეგ კი ალბათ წარმოგიდგენიათ რამდენად გაუსაძლისი გახდა ცხოვრება ახალგაზრდა ებრაელისთვის ნაცისტური გერმანიის მიერ ოკუპირებულ ქვეყანაში. იგი მალევე მიიწვიეს ამერიკის შეერთებულ შტატებში, სადაც შეგვიძლია ვთქვამ რომ მან გაასწრო ნაცისტურ რეჟიმს, რადგან სამწუხაროდ ავსტრიაში დარჩენილი მისი ოჯახის ყველა წევრი დაიღუპა. Აბრაჰამი შტატებში კოლუმბიის უნივერსიტეტში დასაქმდა სადაც ცხოვრების ბოლომდე იმუშავა. 

1941 წელს უნივერსიტეტში ვალდის მეთაურობით შეიქმნა სტატისტიკოს მკვლევართა ჯგუფი, რომელიც ამერიკის არმიას უნდა დახმარებოდა ბომბდამშენების ჯავშნების ოპტიმიზაციაში. პრობლემა წარმოადგენდა შემდეგს: გერმანული სახმელეთო თავდაცვა გასაოცარი წარმატებით უმკლავდებოდა ამერიკულ თვითმფრინავებს. ვალდს და მის ჯგუფს ამერიკის არიამ მიაწოდა მონაცემები იმის შესახებ თუ სად ქონდათ ყველაზე ხშირად მოხვედრილი ტყვიები გადარჩენილ ბომბდამშენებს.

![1.%20Wald.png](attachment:1.%20Wald.png)

მონაცემებიდან აშკარა იყო რომ ყველაზე მოწყვლადი ადგილები თვითმფრინავის უკანა ნაწილი და ფრთები იყო. დასკვნაც თითქოს მარტივი გამოსატანი ჩანდა, დაემატებინათ ჯავშანი აღნიშნულ ადგილებში. თუმცა აბრაჰამ ვალდმა, სრულიად საპირისპიროდ, ავიაციას მოუწოდა ძრავის გარშემო ადგილი გაეძლიერებიდან. მისი დასკვნა ეფუძნებოდა იმ ფაქტს, რომ ყველა მონაცემი რაც გააჩნდათ იყო ზუსტადაც რომ გადარჩენილი თვითმფრინავების და არა ჩამოვარდნილების. შესაბამისად მისი დაშვებით, თვითმფირანვების ჩამოვარდნის ყველაზე ხშირი მიზეზი ძრავში მორტყმული ტყვიები იყო, რასაც შემდეგ „გადარჩენილების მიკერძოებულობა“ ეწოდა და დღემდე ბევრ სხვადასხვა სფეროში გამოიყენება შეცდომების თავიდან ასარიდებლად. სხვა სიტყვებით რომ ვთქვათ მონაცემებიდან დასკვნების გამოტანამდე, დარწმუნებულებუ უნდა ვიყოთ, რომ ის **არ არის** მიკერძოებული ერთი ტიპის ინფორმაციის მოწოდებაზე.

### 1.2 Რა არის მონაცემთა ანალიზი და რატომ დღეს 

იმედი გვაქვს Წინა თავში მოყვანილმა ისტორიულმა მაგალითმა მკითხველს გაუღვივა მონაცემების ანალიზისადმი ინტერესი, თუმცა სანამ უშუალოდ მეთოდების აღწერასა და შესწავლაზე ვისაუბრებთ, ჯერ კარგად განვმარტოთ რას ნიშნავს მონაცემთა ანალიზი და რატომ არის დღეს ასეთი აქტუალური.

**მონაცემების მოგროვების, გასუფთავების, კვლევისა და მოდელირების პროცესს, რომელიც ემსახურება ახალი მიგნებების პოვნას და გვეხმარება გადაწყვეტილებების მიღებაში, მონაცემთა ანალიზი ეწოდება. სხვა სიტყვებით, რომ ვთქვათ: ის გვეხმარება გაურკვევლობის შემცირებაში.**

დისციპლინა რა თქმა უნდა არ არის ახალი, უკვე საუკუნეებია მონაცემთა ანალიზი გამოიყენება სხვადასხვა სფეროში გადაწყვეტილებების მისაღებად. თუმცა დღეს ის რადიკალურად განსხვავებულია. კომპიუტერების გაჩენასთან ერთად ექსპონენციალურად იზრდება მონაცემების დამუშვების სისწრაფეც და მასზე ხელმისაწვდომობაც.

Მაგალითისთვის სურათზე ხედავთ როგორ ხდებოდა მონაცემების მოგროვება (4 მილიონი გამოყენებული ბილეთი) 1939 წელს ლონდონის მეტროს რიგების გასაანალიზებლად.

![2.%20london.png](attachment:2.%20london.png)

ეს კი მთლიანი პროცესის მხოლოდ მცირედი ნაწილია. ამას Შემდგომ მოსდევს: მონაცემების სწორად აღრიცხვა, დახარისხება და დამუშავება. ყველაფერი ეს კი მხოლოდ კალმის და ფურცლის გამოყენებით. 

დღეს მთელი ეს პროცესი გაციფრულებულია: მყისიერად აისახება ყველა ბილეთის შეძენა თუ მრავალჯერადი აბონიმენტის გამოყენება, აწყობილია მზა ანალიტიკური დეშბორდები სადაც შედეგი აისახება და შემდგომ ხდება ამ ანალიზე დაფუძნებით გარკვეული გადაწყვეტილებების მიღება.

Პროცესების, დამუშავების და ხელმისაწვდომობის მკვეთრი გაუმჯობესება, როგორც უკვე აღვნიშნეთ თავიდან ბოლომდე კომპიუტერული ტექნოლოგიების დამსახურებაა.

##### ერთი კომპიუტერი = ერთი პროგრამა = ერთი მომხმარებელი

ალბათ ყველას გინახავთ, როგორ გამოიყურება პირველი კომპიუტერი. გაურკვეველი, უშველებელი მოწყობილობა, რომლესაც მხოლოდ ტრივიალური ამოცანების გადაჭრა შეუძლია და მის გამოსაყენებლად დოქტორის ხარისხიც კი არ კმარა იმდენად კომპლექსურია.

##### ერთი კომპიუტერი = N პროგრამა = M მომხმარებელი

ტექნოლოგიის განვითარებასთან ერთად ახალი შესაძლებლობები გაჩნდა. ერთ კომპიუტერზე უკვე რამდენიმე პროგრამა ყენდებოდა, თუმცა თავად პროგრამები იმდენად რთულად გამოსაყენებელი იყო, რომ  ისევ თითოეულ მათგანს თავის პროფესიონალი სჭრიდებოდა. Სიჩქარეც რა თქმა უნდა დაიხვეწა თუმცა ჯერ კიდევ შორს იყო დღევანდელი რეალობისგან.

##### ერთი კომპიუტერი = N პროგრამა = 1 მომხმარებელი
განვითარების შემდეგ ეტაპზე, უკვე ისეთი კომპიუტერები მივიღეთ, რომლითაც ახლა თქვენ ამ წიგნს კითხულობთ (თუ ნაბეჭდ ვერსიას კითხულობთ მაშინ გეტყვით, რომ ასეთ კომპიუტერში დაიბეჭდა ეს წიგნი). Პროგრამები იმდენად დაიხვეწა, რომ დღეს არ არსებობს ადამიანი ვისაც კომპიტერთან შეხება ქონია და რამდენიმე მათგანს თავისუფლად არ იყენებს. Მაგალითდ ვერ ბრაუზერი, ვორდი, ექსელი, ფოტოების დასათვალიერებელი პროგრამები და ა.შ. 

##### M კომპიუტერი = N პროგრამა = 1 მომხმარებელი

Მიუხედავად იმისა, რომ ყოველდღიურ ცხოვრებაში ჯერ კიდევ ლოკალურ კომპიუტერებს მოვიხმართ, გამოთვლითი ტექნოლოგიების თვალსაზრისით სურათი აქ რადიკალურად შეცლილია. კერძოდ, ქლაუდ ტექნოლოგიების დახმარებით დღეს უკვე შესაძლებელია რამდენიმე სუპერ მძლავრ კომპიუტერს ერთდროულად ვიყენებდეთ, სადაც ასევე რამდენიმე პროგრამა იქნება დაყენებული, ეს ყველაფერი კი მხოლოდ ერთმა (არა მაინცდამაინც დოქტორის ხარისხის და გენიალური ნიჭის მქონე ადამიანმა, როგორც ეს საწყის ფაზაში იყო) ადამიანმა დაბრკოლებების გარეშე აკეთოს.

ყველა ამ ცვლილებამ მიგვიყვანა იქამდე, რომ ნებისმიერ ადამიანს შეუძლია მილიონობით ჩანაწერიან მონაცემებთან ისეთი სისწრაფით მუშაობა, როგორითაც ადრე 2+2 ოპერაცია ითვლებოდა. Ასევე აღსანიშნავია ის ფაქტიც რომ დღეს მონაცემების შეგროვება ხდება ყოველ ფეხის ნაბიჯზე: სოციალური ქსელი, საბანკო ბარათის ტრანზაქცია, ინტერნეტ ტელევიზია, სხვადასხვა საიტების გამოყენება და ა.შ. Ზუსტად ამიტომ არის დღეს ასეთი აქტუალური მონაცემების დამუშავების მეთოდების ცოდნა, ის ჩვენი ყოველდღიურობაა.

### 1.3 მონაცემთა ანალიზის ტიპები
ამ ეტაპზე ისტორიული გადახვევები საკმარისია, დროა გადავიდეთ საქმეზე და გავეცნოთ ანალიზის ტიპებს, რომელებიც დღეს გამოიყენება. Სხვადასხვა ლიტერატურასა თუ სტატიაში შეხვდებით განსხვავებულ დაყოფას, ჩვენთვის ყველაზე ახლოს გარტნერის განმარტებაა, სადაც ოთხი ძირითადი ნაწილია, დალაგებული მათი სირთულისა და კომპლექსურობის მიხედვით:

#### 1.3.1 აღწერითი ანალიზი

აღწერითი ანალიზი პასუხობს კითხვას - **რა მოხდა?** მისი გამოყენებით ვაანალიზებთ მონაცემებს სხვადასხვა ფაქტორებისა და დროის ჭრილებში. ისტორიული მეტრიკები რომლებზე დაკვირვებაც შეიძლება გვაინტერესებდეს მოიცავს ისეთ მრავალფერვან სპექტრს როგორებიცაა: საშუალო ჩეკის მოცულობა ყოველთვიურად, კოპმანიის მოგება, მომხმარებლების განაწილება ასაკისა და შეძენილი პროდუქტების მიხედვით და ა.შ. Რა თქმა უნდა ეს პროცესი გვეხმარება გამოვიტანოთ დასკვნები, რომლებზე დაყრდნობითაც სამომავლოდ უკეთ წარვმართავთ საქმეს. სხვა სიტყვებით რომ ვთქვათ აღწერითი ანალიზი არის ისტორიული მონაცემების ინტერპრეტაციის პროცესი. Ნებისმიერი ანალიტიკური ამოცანისთვის, აღწერითი ანალიზი არის საწყისი და აუცილებელი ნაბიჯი სწორი გადაწყვეტილებების მისაღებად, თუმცა ხშირ შემთხვევაში ეს საკმარისი არ არის.

#### 1.3.2 დიაგნოსტიკური ანალიზი

დიაგნოსტიკური ანალიზი პასუხოს კითხვას - **რატომ მოხდა?** Შეგვიძლია ვთქვათ, რომ ის აღწერითი ანალიზის ლოგიკური გაგრძელებაა. Მრავალი სტატისტიკური თუ ალბათური მეთოდის არსებობის მიუხედავად, დიაგნოსტიკური ანალიზის განუყოფელი ნაწილია შინაარსობრივი კვლევა. Რა თქმა უნდა ყველაფერი რაოდენობრივით იწყება როგორებიცაა: ჰიპოთეზების ჩამოყალიბება და შემოწმება, ცვლადებს შორის დამოკიდებულების და მიზეზ შედეგობრიობის დადგენა (ძალიან ხშირად ფიქრობენ, რომ კორელაცია მიზეზ შედეგობრიობასაც ნიშნავს რამაც შეიძლება მნიშნველოვან შეცდომებამდე მიგვიყვანოს), ასევე შეიძლება მაპროგნოზირებელი მოდელებიც (რომელზეც შემდეგ ქვეთავში ვისაუბრებთ) გამოვიყენოთ ხდომილებების ასახსნელად. თუმცა ამ ყველაფერთან ერთად აუცილებელია ამოცანის შინაარსის სიღმისეული ცდონა, გამოცდილება და ალღო იმისათვის რომ მონაცემებზე დაყრდნობით ავხსნათ მოვლენები. ერთის მხრივ კომპლექსურობა ანალიზს ართულებს თუმცა ამავედ დროს მეტ თავდაჯერებულობას გვმატებს გადაწყვეტილებების მიღებაში. Როგორც თავის დასაწყისში აღვნიშნეთ, მონაცემთა ანალიზის არსიც ხომ ისაა რომ მომავლის გაურკვევლობა შეამციროს, ანუ მომავალი გვაინტერესებს და არა წარსული. აღწერითი და დიაგნოსტიკური ანალიზის მთავარი მიზანიც (მაგრამ არა ერთადერთი), წარსულ ხდომილებებზე დაყრდნობით მომავლის მეტი სიზუსტით პროგრნოზირებაა.

#### 1.3.3 მაპროგნოზირებელი ანალიზი

მაპროგნოზირებელი ანალიზი პასუხობს კითხვას - **რა მოხდება?** აღნიშნულზე პასუხის წარმატებით გასცემად, ზემოთ მოყვანილი ანალიზის ორივე ტიპი აუცილებელი წინაპირობაა სასურველი შედეგის მისაღებად. თუმცა ხშირ შემთხვევაში საკმარისი ნამდვილად არ არის რისთვისაც მაპროგნოზირებელ ანალიზს ვიყენებთ. Მიზნის მისაღწევად, Მთელი რიგი მოდელები და იარაღებია. ეს წიგნიც სწორადაც ამ და ნაწილობრივ ანალიზის შემდეგ ტიპზე კონცენტრირდება.

მომავალ თავებში დეტალურად განვიხილავთ რა ტიპის მოდელები გამოიყენება სხვადასხვა ხდომილებების საპროგნზოდ. კითხვა აქ არ გაწყვიტოთ, რა თქმა უნდა მხოლოდ საუბრით არ შემოვიფაგლებით, პირიქით ყველაზე მეტ დროს მანქანური დასწავლის ალგორითმების რეალურ ცხოვრებაში გამოყენების შესწავლას დავუთმობთ.

#### 1.3.4 რეცეპტული (prescriptive) ანალიზი

რეცეპტული ანალიზი პასუხობს კითხვას - **რა მოვიმოქმედოთ შემდგომ?** გარკუველი (სასურველი) ალბათობით გქონდეს ხედვა იმაზე თუ რას მოელი მომავალში ფასდაუდებელი ინფორმაციაა, მაგრამ ზოგიერთ შემთხვევაში ეს პირდაპირ არ ითარგმნება იმაში თუ რა ქმედებების გაწევაა ოპტიმალური. Სწორად ამისთვის გამოიყენება რეცეპტული ანალიზი. Სხვა სიტყვებით, რომ ვთქვათ ეს არის პროცესი, რომელიც გვეხმარება შესაძლო სცენარების, არსებული რესურსების, წარსული გამოცდილებისა და მომავლის პროგნოზის გათვალისწინებით მივიღოთ ოპტიმალური გადაწყვეტილებები.

უკანასკნელი ორი ანალიზის ტიპი მჭიდროდ გადაჭვულია. ხშირად მათ ერთმანეთშიც ურევენ ან აიგივენებ. ამიტომ Შევეცდებით მაგალითების საფუძველზე უფრო მკაფიოდ ავხსნათ მათ შორის განსხვავებები.
Პირველ მაგალითად განვიხილოთ საკრედიტო ბარათების თაღლითობა. Რა თქმა უნდა აქ ამოცანის პირველი ნაწილია წინასწარ გამოვიცნოთ კონკრეტული მომხმარებელების თაღლითობები, რასაც მაპროგონიზრებელი ანალიზის გარკვეული ალგორითმებით ვუპასუხებდით. თუმცა ამოცანა აქ არ სრულდება. Რას ვშვრებით მას მერე რაც ვიცით რომ გარკვეული მომხმარებელები დიდი ალბათობით თაღლითობენ? ყველას ავტომატურად ვუბლოკავთ საკრედიტო ბარათს? კი მაგრამ ხომ ვიცით რომ არ არსებობს 100%-იანი სიზუსტის მოდელი და ბევრ ადამიანს ტყუილად რომ დავუბლოკოთ ბარათზე წვდომა რეპუტაციული რისკიც გასათვალისწინებელია. იქნებ აზრიანია თაღლითობის ალბათობასთან ერთად მის სავარაუდო მოცულობებსაც შევხედოთ? ან ისე ხომ არ არის საქმე, რომ მოდელი რატომღაც ყველა დიდი მოცულობის ტრანზაქციას თაღლითობად აფასებს (შეიძლება ეს 90% შემთხვევაში ასეც იყოს) და მდიდარი კლეინტები ხომ არ უნდა გამოვხშიროთ ბარათის ავტომატური დაბლოკვის პროცესისგან? Მოკლედ რომ ვთქვათ პროგრნოზიდან ოპტიმალური გადაწყვეტილების მიღებამდე კიდევ დიდი გზაა გასავლელი, ბევრი ფაქტორი გასაანალიზებელი და ბევრი სცენარი განსახილველი. ამ ყველაფერს კი ზუსტადაც რეცეპტული ანალიზი (Prescriptive Analytics) კრავს. ანუ მისი დახმარებით საბოლოო ჯამში ვიღებთ ოპტიმალური ქმედებების რეკომენდაციებს.

ასევე კარგი მაგალითია თავად რეკომენდაციების სისტემები. ტერმინი შეიძლება ახალია, მაგრამ მას ყველგან ხვდებით: ვიდეოს დასრულებისას youtube თქვენ წარსულ ნახვებსა და გემოვნებაზე დაყრდნობით გირჩევთ შემდეგ ვიდეოებს, ფეისბუქზე ახალი მეგორების დამატების შემოთავაზებებსაც შეხვდებოდით, spotify-ის მომხმარებელი თუ ხართ ხომ საერთოდ ბედნიერი იქნებით ამ ფუნქციონალით. Საკმაოდ მძლავრი და კოპლექსური სისტემებია, რაც საკუთარ თავში მოიცავს როგორც მანქანური დასწავლის ალგორითმს, ასევე მომდევნო ოპტიმალური ქმედებას (ის ხომ თავისით ადგენს რა მოგეწონებათ და მერე გთავაზობთ კიდეც). 

Იმედი გვაქვს ზემოთ მოყვანილი მაგალითები დაგეხმარათ უკეთ გაგეანალიზებინათ თუ რა სხვაობაა მაპროგნოზირებელ და რეცეპტულ ანალიზის ტიპებს შორის.

შეჯამებისთვის გთავაზობთ გარტნერის გრაფიკს და ჩვენს განმარტებებს თუ რა დარგები პასუხობენ შესაბამის ანალიზის ტიპებს:

![3.%20analtics%20types.png](attachment:3.%20analtics%20types.png)

### 1.4 მონაცემთა მეცნიერება და მანქანური დასწავლის ალგორითმების ტიპები

#### 1.4.1 მონაცემთა მეცნიერება
ახლა კი მოდით წინა აბზაცებში დაპირებულ, ყველაზე საინტერესო, ნაწილზე გადავიდეთ. Ბოლოსდაბოლოს რა ჯადოქრობაა ეს მონაცემთა მეცნიერება, ან საერთოდ მეცნიერებაა? არა მეცნიერება ნამდვილად არ არის, **ის არის მულტიდისციპლინარულ ცოდნაზე დაფუძნებული მეთოდოლოგია**. Შესაბამისად არც მონაცემთა მეცნიერები არიან ჯადოქრები. ისინი უბრალოდ უკეთესი პროგრამისტები/დეველოპერები არიან ვიდრე სტატისტიკოსები და უკეთესი სტატისტიკოსები ვიდრე პროგრამისტები/დეველოპერები. ამას დაკნინებად ნუ გაიგებთ, Მარტივად კი ჟღერს, თუმცა არც ისეთი მარტივი პროფესიაა, რომელიც საკმაოდ სიღმისეულ ცოდნას მოითხოვს სტატისტიკის, მათემატიკურ მოდელირების, კომპიუტერული მეცნიერებების და ბიზნეს ანალიზის.

#### 1.4.2 მანქანური დასწავლის ალგორითმის ტიპები
სავარაუდოდ ერთი სული გაქვთ როდის გადავალთ პრაქტიკულ ნაწილზე, თუმცა ჯერ ძალიან ზოგადად, შეგრძნების დონეზე, მოგიყვებით თუ რა ტიპის ამოცანები იჭრება მანქანური დასწავლით და რა ტიპის ალგორითმები შეესაბამება მათ. Შემდგომ თავებში კი ამ ყველაფერს უფრო სიღრმისეულად, პრაქტიკული სავარჯიშოებით და ამოცანებით გავივლით.

Წიგნის  მიზანი და  ამბიცია ნამდვილად არ არის მკითხველი მონაცემთა მეცნიერად აქციოს. Შესაბამისად ჩვენ მანაქანური დასწავლის მხოლოდ ძირითად ნაწილებს შევეხებით. არ გეგონოთ, რომ რამეს გაკლებთ. ეს ძირითდი ნაწილები რეალური სამყაროს, მეტს თუ არა, 80%-ს ნამდვილად ფარავს.

მანქანური დასწავლის ალგორითმები შეგვიძლია 4 მსხვილ მიმართულებად დავყოთ: დაკვირვებითი მოდელები, რაც თავის მხრივ რეგრესიულ და კლასიფიკაციის ტიპებად იშლება, და დაკვირვების გარეშე მოდელები, რომელიც კლასტერირებად და ასოციაციად იშლება.

![4.%20algorithms.png](attachment:4.%20algorithms.png)

##### 1.4.2.1 დაკვირვებითი ალგორითმები
დაკვირვებითი მოდელები შეიძლება ითქვას, რომ ყველაზე ინტუიტიურია მანქანურ დასწავლაში. Მოდელს ვაწვდით მონაცემებს მათი შესაბამისი პასუხებით. Შემდგომ კი იგი ცდილობს იპოვნოს ისეთი კანონზომიერებები, რაც მიწოდენული ინფორმაციით სწორი პასუხების პოვნაში დაეხმარება. გამოდის, რომ კომპიუტერს ბევრი კარგი მაგალითის საფუძველზე ვასწავლით ამოცანის ამოხსნას. 

Მაგალითისთვის გვინდა გამოვიცნოთ დაიგვიანებს თუ არა თვიმფმრინავის რეისი. ამოცანის გადასაჭრელად დაგვჭირდება თითოეულ რეისთან დაკავშირებული წარსული ინფორმაცია: გამოფრენის დრო, გამოფრენის ქალაქი, დანიშნულების ქალაქი, ამინდი გამოფრენის ქალაქიდან, ამინდი დანიშნულების ადგილზე, ავია კომპანიის სახელი და ა.შ. (ამ ცვლადების ერთობლიობას ბევრი სახელით მოიხსენიებენ: **დამოკიდებული ცვლადები, მახასიათებელი ცვალდები, მოდელში შემავალი ცვლადები**) Ხოლო პასუხების მხარეს (**პსუხებს კი: დამოკიდებულ ცვლადებად ან საპროგნოზო ცვალდებად მოიხსენიებენ**) იქნება დაიგვიანა/დროზე მოფრინდა. Მოდელი იპოვნის კანონზომიერებას დამოკიდებულ ცვლადებსა და დაგვიანებას შორის, რის შემდეგაც უკვე მხოლოდ მახასაითებელი ცვლადების საფუძველზე გააკეთებს პროგნოზს: დააგვიანებს რეისი თუ დროზე მოფრინდება. 

Ზემოთ მოყვანილი მაგალითი კლასიკური **კლასიფიკაციის** ამოცანაა. Მსგავსი ტიპის ამოცანის **მიზანი ორ ან რამდეინიმე კლასს (კატეგორიას) შორის სწორი პასუხის პროგნოზირებაა.** ჩვენი მაგალითის მიზანი, რომ ყოფილიყო გამოგვეცნო ზუსტად რამდენ წუთს დაიგვიანებდა რეისი, რეგრესიული ამოცანა იქნებოდა. **ანუ რეგრესიული ტიპის მოდელები პროგრნოზირებენ უწყვეტ მნიშვნელობებს.**

##### 1.4.2.2 დაკვირვების გარეშე ალგორითმები
დაკვრივების გარეშე ალგორითმები შედარებით აბსტრაქტულია. აქ არც ჩვენ და არც კომპიუტერმა “სწორი” პასუხები არ ვიცით. შესაბამისად ალგორითმებიც აღწერითი ხასიათისაა და დაკვირვების გარეშე ფუნქციონირებენ. კლასტერირების შემთხვევაში (რომელიც ყველაზე ხშირად გამოყენებადი და გავრცელებულია) ალგორითმის მიზანი, მრავალცვლადიანი მონაცემების გარკვეული **მსგავსების მიხედვით დაჯგუფებაა.**

Კლიენტების სეგმენტაცია Სხვა მრავალ ამოცანასთან ერთად წარმოადგენს, ერთ-ერთი ყველაზე გავრცელებული ამოცანას, რომელიც კლასტერირების ალგორითმებით იჭრება. Სეგმენტაცია შეიძლება იყოს დემოგრაფიული, ქცევითი, გეოგრაფიული და ა.შ. ცვალდებით. Მიზანი კი ერთგვაროვანი კლიენტების დაჯგუფება და მათზე მეტად პერსონიზირებული კამპანიების წარმოებაა. 

კლასტერირების გარდა დაკვირვების გარეშე მოდელების ყველაზე მსხვილი მიმართულება ასოციაციის ალგორითმებია. Სავარაუდოდ მკითხველების უმეტესობას შეხვედრია ამაზონზე პროდუქტის გვედრზე ყოფნისას შემდეგი ჩანაწერი: “ამ პროდქუტთან ერთად ყველაზე ხშირად X პროდუქტს ყიდულობენ მომხმარებლები”. 

![5.%20amazon.jpeg](attachment:5.%20amazon.jpeg)

ან შესაძლებელია ლუდი და “პამპერსის” ლეგენდაც გსმენიათ (პამპერსის საყიდლად წასული ახალგაზრდა მამები ხშირად ყიდულობნენ ლუდსაც, რის აღმოჩენის შემდეგაც მათი მაღაზიაში ერთმანეთის გვერდით განთავსება გადაწყვიტეს გაყიდვების გასაზრდელად). Სხვა სიტყვებით, რომ ვთქვათ **ასოციაციის ალგორითმი ეძებს ისეთ პროდუქტებს რომელებიც ერთამნეთის გაყიდვას უწყობენ ხელს / ერთად ყველაზე ხშირად იყიდებიან.**  

აქვე მცირედი გადახვევა თუ რა ღირებულების მომტანია რეალურ გარემოში: ასოციაციის ალგორითმები ხშირად რეკომენდაციების სისტემებში გამოიყენება. Რეკომენდაციების სისტემები კი (რაზეც მოგვიანებით ცალკე თავიც იქნება მიძღვნილი) სხვადასხვა შეფასებებებით “ამაზონის” შემოსალვების 35% აგენერირებენ.

### 1.5 Მონაცემებზე დაფუძნებული ამოცანის ციკლი

ყველაფერი რაც აქამდე ვისაუბრეთ, ერთი დიდი ციკლის, მონაცემებზე დაფუძნებული ამოცანის, ნაწილია. Ხშირად გვიჩნდება ცდუნება, რომ  ამოცანის გადაჭრა პირადპირ ალგორითმების შერჩევით დავიწყოთ და ვნახოთ შედეგი. თუმცა სწორი გზა რა თქმა უნდა ამოცანიის შინაარსობრივი გააზრებით დაწყებაა. პრობლემის პირველად გაცნობას ყოველთვის მოსდევს მონაცემების კვლევა. Პროექტის ამ ეტაპზე აუცილებლად ჩნდება დამატებითი კითხვები ამოცანის შინაარსთან დაკავშირებით და მასთან მიბრუნება მთლიანი პროცესის მიკრო ციკლია.

![6.%20data%20Cycle.png](attachment:6.%20data%20Cycle.png)

Მოდელისთვის მონაცემების მომზადება (რასაც ცალკე ქვეთავებად ვისწავლით წიგნის შემდგომ ნაწილებში) იწყება მხოლოდ მას შემდეგ, რაც ამოცანის შინაარსთან და მონაცემებთან საკამრისი კომფორტი გაგვაჩნია. ეს კი მოდელირებასთან ერთად  მთლიანი პროცესის კიდევ ერთი მიკრო ციკლია.

Პირველი მოდელის რეზუტლტატი თითქმის არასდროს არის სახარბიელო. შესაბამისად მისი შედეგის გაანალიზებისას ხშირად ხდება ზუსტადაც ამოცანის საწყის ეტაპზე, შინაარსობრივ გააზრებაზე, დაბრუნება. სასურველი რეზულტატი მხოლოდ ამ ციკლის რამდენჯერმე და საფუძვლიანად განმეორების შედეგად დგება. მხოლოდ მთელი ამ პროცესის რამდენჯერმე განმეორება გვაძლევს მოდელის სიზუსტესა და მდგრადობაში დარწმუნების საფუძველს, რასაც პროცესის ბოლო საფეხური, მოდელის რეალურ გარემოში დანერგვა, მოსდევს.

### 1.6 პირველი შეხება, “მაგიურ”, მანქანურ დასწავლასთან
თავის ყველაზე საინტერესო, პრაქტიკულ, ნაწილთანაც მოვედით. აქ თქვენივე ხელებით შეძლებთ, ორი არც თუ ისე მარტივი ამოცანის გადაჭრას და დარწმუნდებით, რომ მანქანური დასწავლა მართლაც არ არის ჯადოქრობა!

დღეს ნამდვილად აღარ არის მანქანური დასწავლის ალგორითმების გამოყენება ისეთი რთული, როგორც ის 5 და 10 წლის წინ იყო. თუმცა ისეთი მარტივიც არ არის, რომ პირველსავე თავში წარმატებით შევძლოთ მისი გამოყენება. სწორად ამიტომ დახმარებისთვის ავტომატური მანქანური დასწავლის (AutoML) ბიბლიოთეკას მივმართავთ.

**ავტომატური მანქანური დასწავლა (AutoML) არის რეალური პრობლემების მანქანური დასწავლის პროცესების გამოყენებით გადაჭრის ავტომატიზაცია.** მისი გამოყენებისთვის ყოველთვის ნამდვილად არ არის საჭირო იყოთ მონაცემთა მეცნიერი ან კარგად გესმოდეთ კოდის წერა. ხშირ შემთხვევაში ასეთ პლატფორმებს ძალიან ინტუიტიური და მარტივად გამოსაყენებელი ინტერფეისები აქვთ. თუმცა გამომდინარე იქედან, რომ ყველა ასეთი პლატფორმა ფასიანია, როგორც ზემოთ აღვნიშნეთ ჩვენ AutoML-ის, უფასო ბიბლიოთეკას, და არა პლატფორმას, გამოვიყენებთ.

*შენიშვნა: თუ თქვენ ამ კოდის მხოლოდ წაკითხვა არ გაკმაყოფილებთ (რისი ძალიან დიდი იმედიც გვაქვს) და გსურთ მისი გაშვება, ამისათვის თქვენს კომპიუტერზე Jupyter Notebook-ის დაყენე დაგჭირდებათ (გირჩევთ Anaconda დააყენოთ, სადაც Jupyter-იც იქნება) ან კიდევ უფრო მარტივად Google-ის ანგარიში და Google Colab-ზე ამ ნოუთბუქის Github-ით იმპორტიც საკმარისი იქნება.*

ცხრილურ მონაცემებთა მომუშავე ერთ-ერთი ასეთი ავტომატიზირებული მანქანური დასწავლის ბიბლიოთეკა არის **autogluon**. ეს გენიალური ბიბლითეკა AWS-ის გუნდის მიერაა შემუშავებული და მისი დახმარებით სულ რაღაც 5-10 ხაზიანი კოდის დაწრითაა შესაძლებელი სასურველი შედეგების მიღება.

*შენიშვნა: ამ ქვეთავში ბევრ ისეთ ბრძანებას შეხვდებით, რომელიც მკითხველების უმეტესობისთვის უცხო იქნება. ამან არ შეგაშინოთ, წიგნის მომდევნო თავებში ყველაფერს დეტალურად დავფარავთ*

აღნიშნული ბიბლიოთეკის გამოსაყენებლად პირველ რიგში მისი დაყენება დაგვჭირდება, რაც შემდეგი ბრძანებებითაა შესაძლებელი:

In [ ]:
!pip install -U pip
!pip install -U setuptools wheel
!pip install -U "mxnet<2.0.0"
!pip install autogluon

მას შემდეგ რაც დავაყენეთ ბიბლიოთეკა, აუცილებელია მისი სამუშაო გარემოში შემოტანა შემდეგი ბრძანებებით:

In [ ]:
import pandas as pd #ბიბლიოთეკა ცხრილების წასაკითხად 
from autogluon.tabular import TabularDataset, TabularPredictor #autogluon–ს ცხრილებზე სამუშაო ბიბლიოთეკა
from sklearn.model_selection import train_test_split #მონაცემების სატესტო და საწვრთნელ ნაწილებად გახლეჩის ბიბლიოთეკა

#### 1.6.1 კლასიფიკაციის ამოცანა

როგორც წინა აბზაცებში აღვნიშნეთ, ამოცანის გადაჭრა მისი შინაარსობრივი გააზრებით იწყება. სწორად ამიტომ ძალიან მოკლედ და მარტივად აღვწეროთ რა პობლემას ვეჭიდებით: ამოცანის მიზანია გარკვეული ინფორმაციით თითოეულ ადამიანზე ვიპროგნოზოთ გადარჩებოდა თუ არა ის ტიტანის ტრაგედისას:

   * PassengerId - მგზავრის ID
   * Pclass - კაბინის კლასი
   * Name - მგზავრის სახელი
   * Sex - სქესი
   * Age - ასაკი
   * SibSp - დების/ძმების/მეუღლის რაოდენობა გემზე
   * Parch - მშობლების/შვილების რაოდენობა გემზე
   * Ticket - ბილეთის ნომერი
   * Fare - ბილეთის ფასი
   * Cabin - კაბინის სახელი
   * Embarked - ჩასხდომის პორტი
   
სხვა სიტყვებით რომ ვთქვათ, თითოეულ ადამიანზე მოდელს ვაწვდით ამ ინფორმაციას. ასევე ვაწვდით გადარჩა თუ არა ადამიანი და ვავალებთ კანონზომიერებების პოვნას. რათა შემდგომ მხოლოდ ზემოთ ჩამოთვილი ინფორმაციით იპროგნოზოს გადარჩება თუ არა ადამიანი.    
   
პასუხები კი **Survived** ცვალდშია მითითებული, სადაც 1-ნიშნავს გარადჩენას, 0-დაღუპვას.

ამ მონაცემების სამუშაო გერემოში შემოსატანა და მისი პირველი 3 ჩანაწერის გამოსატანად კი შემდეგი ბრძანებებია საჭირო:

In [ ]:
df_class = pd.read_csv('titanic.csv') #მონაცემების სამუშაო გარემოში შემოტანა
df_class.head(3) #შემოტანილი მონაცემის პირველი 3 ჩანაწერის გამოჩენა

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

ზოგადად ნებისმიერი მანქანური დასწავლის მოდელირების დაწყებამდე საჭიროა მონაცემების "სავარჯიშო" და "სატესტო" ნაწილებად დახლეჩვა, რათა დარწმუნებულები ვიყოთ მოდელის მიუკერძოებულობაში (რა თქმა უნდა ამაზეც მომდევნო თავებში დეტალურად მოგიყვებით). შესაბამისად შემდეგი ბრძანება მთლიან მონაცემებს გახლეჩს 80%-20% პროპორციით (სავაჯიშო/სასწავლი ნაწილი 80%, სატესტო ნაწილი 20%):

In [ ]:
train_class, test_class = train_test_split(df_class, test_size = 0.2, random_state=50) 

ასევე აუცილებელია შემოვიღოთ რაიმე მეტრიკა, რის დახმარებითად გავზომავთ მოდელის წარმადობას.

In [ ]:
metric_class = 'accuracy'

ვუთხრათ მოდელს, რომელი ის ცვლადი მთელი ცხრილიდან, რის პროგნოზირებაც გვინდა:

In [ ]:
label_class = 'Survived'

გადავიყვანოთ მონაცემები autogluon–თვის მისაღებ ფორმატში:

In [ ]:
train_data_class = TabularDataset(train_class) 
test_data_class = TabularDataset(test_class)

და აი ისიც, კოდი რომელიც ჩვენ პირველ მოდელს გაუშვებს:

In [ ]:
predictor_class = TabularPredictor(label = label_class, eval_metric = metric_class) \
            .fit(train_data_class, time_limit = 300)

# label-ში ვუთითებთ საპროგნოზო ცვლადის სახელს (ზემოთ ჩვენ მასაც label დავარქვით)
# eval_metric-ში ვუთითებთ მოდელის წარმადობის შესაფასებელ მეტრიკას (ზემოთ  metric დავარქვით)
# time_limit - ნიშნავს იმ წამების რაოდენობას რაც გვსურს რომ დაიხარჯოს მოდელის აგებაზე/ოპტიმიზაციაზე.
# რაც უფრო დიდ დროს ვუთითებთ მით მეტია შესაძლებლობა მოდელის სიზუსტის, თუმცა აქ სიამტივისთვის 300 წამიც გვეყოფა

დიახ, ჩვენი პირველი მოდელი უკვე მზადააა. ისღა დაგვრჩენია შევამოწმოდ ჩვენს გადანახულ, სატესტო, მონაცემებზე და ვნახოთ რამდენად კარგად შეძლო დაკისრებული დავალების შესრულება:

*შენიშვნა: ქვემოთ მოცემულ კოდს, ჩვენს მიერ არჩეული მეტრიკის გარდა (accuracy), კიდევ ბევრი სხვა მეტრიკის შედეგიც გამოაქვს, თუმცა ამ ეტაპზე ყურადღებას ნუ მივაქცევთ. მეტრიკებზე სულაც ცალკე თავი გვაქვს მიძღვნილი.*

In [ ]:
predictor_class.evaluate(test_data_class)['accuracy']

Evaluation: accuracy on test data: 0.7988826815642458
Evaluations on test data:
{
    "accuracy": 0.7988826815642458,
    "balanced_accuracy": 0.7750698146737751,
    "mcc": 0.6079964751009972,
    "roc_auc": 0.8805534399593806,
    "f1": 0.71875,
    "precision": 0.92,
    "recall": 0.5897435897435898
}


0.7988826815642458

ვხედავთ, რომ ის მეტრიკა (accuracy) რითიც ვთხოვეთ შეფასება 0.7989-ის ტოლია. ეს კი იმას ნიშნავს რომ 79.89% შემთხვევაში მოდელმა სწორად გამოიცნო გადარჩებოდა თუ არა ადამიანი ტიტანიკის ტრაგედიისას. შეიძლება ითქვას, რომ მშვენიერი რეზულტატია, განსაკუთრებით ასეთი მცირე და ზედაპირული ძალისხმევის შედეგად.

#### 1.6.1 რეგრესიული ამოცანა

დაპირებული, მეორე ამოცანა რეგრესიული ტიპისაა, ანუ ამ შემთხვევაში მოდელს ზუსტი (უწყვეტი) მნიშვნელობის გამოცნობას დავავალებთ: მოცემული გვაქვს სადაზღვევო ხარჯები და იმ ადამიანების შესახებ რამდენიმე ცვლადი ვიზეც მოხდა ამ ხარჯების გაწევა:

   * age - ასაკი
   * sex - სქესი
   * bmi - სხეულის მასის ინდექსი
   * children - დაზღვევაში ჩართული შვილების რაოდენობა
   * smoker - არის თუ არა მწეველი
   * region - საცხოვრებელი რეგიონი
   
ხოლოს საპროგნოზო ცვლადი, როგორც ზემოთ აღვნიშნეთ არის დანახარჯის რაოდენობა - **charges**

*შენიშვნა: მიუხედავად იმისა, რომ ამოცანა არის რეგრესიული ხასიათის, autogluon-ის ერთ-ერთი მთავარი ღირებულება ისაა, რომ თვითონ ხვდება რა ტიპის პრობლემასთან აქვს საქმე. შესაბამისად ის ნაბიჯები რაც მოდელის გაშვების ჩათვლით დაგვჭირდება, იქნება წინა მაგალითის ანალოგიური.*

მონაცემების სამუშაო გარემოში შემოტანა და მისი პირველი 3 ჩანაწერის გამოტანა:

In [ ]:
df_reg = pd.read_csv('insurance.csv') #მონაცემების სამუშაო გარემოში შემოტანა
df_reg.head(3) #შემოტანილი მონაცემის პირველი 3 ჩანაწერის გამოჩენა

age     sex    bmi  children smoker     region     charges
0   19  female  27.90         0    yes  southwest  16884.9240
1   18    male  33.77         1     no  southeast   1725.5523
2   28    male  33.00         3     no  southeast   4449.4620

სავარჯიშო/საწვრთელ და სატესტო მონაცემებად გახლეჩვა:

In [ ]:
train_reg, test_reg = train_test_split(df_reg, test_size = 0.2, random_state = 50) 

In [ ]:
test_reg['charges'].mean() # სატესტო ცხრილის საპრონგნოზო ცვლადის, სადაზღვევო დანახარჯის, საშუალოს გამოტანა, რომელიც
                       # მოდელის შესაფასებლად დაგვჭირდება

13990.39168537313

მთავარი განსხვავება მეტრიკის განსაზღვრისას იქნება, რადგან რეგრესიული და კლასიფიკაციის მოდელების შესაფასებლად რადიკალურად განსხვავებული მიდგომაა საჭირო.

In [ ]:
metric_reg = 'mean_absolute_error' #აბსოლუტური საშუალო გადახრის მეტრიკა

საპროგნოზო ცვლადის მითითება

In [ ]:
label_reg = 'charges'

მონაცემების autogluon–თვის მისაღებ ფორმატში გადაყვანა:

In [ ]:
train_data_reg = TabularDataset(train_reg) 
test_data_reg = TabularDataset(test_reg)

მოდელის სწავლება, კვალვაც მარტივი ერთ-ხაზიანი კოდით ხდება:

In [ ]:
predictor_reg = TabularPredictor(label = label_reg, eval_metric = metric_reg) \
                .fit(train_data_reg, time_limit = 300)

გადანახულ/სატესტო მონაცემებზე მოდელის წარმადობის/სიზუსტის შემოწმება:

*შენიშვნა: ქვემოთ მოცემულ კოდს, ჩვენს მიერ არჩეული მეტრიკის გარდა (mean_absolute_error), კიდევ ბევრი სხვა მეტრიკის შედეგიც გამოაქვს, თუმცა ამ ეტაპზე ყურადღებას ნუ მივაქცევთ. მეტრიკებზე სულაც ცალკე თავი გვაქვს მიძღვნილი.*

In [ ]:
predictor_reg.evaluate(test_data_reg)

Evaluation: mean_absolute_error on test data: -2169.0555191760436
	Note: Scores are always higher_is_better. This metric score can be multiplied by -1 to get the metric value.
Evaluations on test data:
{
    "mean_absolute_error": -2169.0555191760436,
    "root_mean_squared_error": -4371.4560319730645,
    "mean_squared_error": -19109627.83947369,
    "r2": 0.8807821282217947,
    "pearsonr": 0.9388401466394608,
    "median_absolute_error": -988.5785021484376
}


{'mean_absolute_error': -2169.0555191760436,
 'root_mean_squared_error': -4371.4560319730645,
 'mean_squared_error': -19109627.83947369,
 'r2': 0.8807821282217947,
 'pearsonr': 0.9388401466394608,
 'median_absolute_error': -988.5785021484376}

მიღებული შედეგებით ვხედავთ რომ საშუალო აბსოლუტური ცდომილება არის 2169. აღნიშნული მეტრიკა, პროგრნოზირებულ და რეალურ მნიშნვლნელობებს შორის საშუალო აბსოლუტურ გადახრას გვიჩვენებს. მისი მნიშნველობის ინტერპრეტაციის გასამარტივებლად, შეგვიძლია მიღებული შედეგი საპროგნოზო ცვლადის რეალური საშუალოთი დავანორმალიზოთ (იმ საშუალოთ ზემოთ რომ დავთვალეთ): 1 – (2169 / 13990) = 0.845, რაც უხეშად, მოდელის, 84.5%–იან სიზუსტეში ითარგმნება.

ზემოთ მოცემული ორივე მაგალითი ზედაპირული და შეიძლება ითქვას საილუსტრაციო ხასითისაა. გვიდოდა გვეჩვენებინა, რომ ზოგ შემთხვევაში მართლაც რამდენიმე ხაზიანი კოდითაც არის შესაძლებელი დამაკმაყოფილებელი შედეგის მიღება. რა თქმა უნდა სწორი იქნებოდა ჯერ აღწერითი და დიაგნოსტიკური ანალიზი ჩაგვეტარებინა, მოცემულ ინფორმაციაზე დაყრდნობით გამოგვეყვანა ახალი ცვლადები, მოგვეძიებინა დამატებითი ინფორმაცია და კიდევ მრავალი ნაბიჯი გაგვევლო ოპტიმალური შედეგის მისაღებად. ყველა ამ ნაბიჯს კი დეტალურად მომდევნო თავებში გავივლით. ჩვენი მიზანიც ეს არის, რომ წიგნის ბოლოს თქვენ სრულყოფილად შეძლოთ მონაცემებზე დაფუძნებული ამოცანების დამოუკიდებლად გადაჭრა.